llvm/lib/Transforms/IPO/AttributorAttributes.cpp

```diff
@@ -12987,6 +12989,225 @@ struct AAAllocationInfoCallSiteArgument : AAAllocationInfoImpl {
+
+  return AAMem && NoAlias && AAMem->isAssumedWriteOnly() &&
+         NoAlias->isAssumedNoAlias() && !Arg.hasPointeeInMemoryValueAttr();
+}
+
+struct AAConvertOutArgumentFunction final : AAConvertOutArgument {
+  AAConvertOutArgumentFunction(const IRPosition &IRP, Attributor &A)
+      : AAConvertOutArgument(IRP, A) {}
+
+  SmallVector<bool> ArgumentsStates;
+
+  /// See AbstractAttribute::updateImpl(...).
+  void initialize(Attributor &A) override {
+    const Function *F = getAssociatedFunction();
+    if (!F || F->isDeclaration())
+      return;
+
+    // Assume that all  args are convertable at the begining.
+    ArgumentsStates.resize(F->arg_size(), true);
+  }
+
+  /// See AbstractAttribute::updateImpl(...).
+  ChangeStatus updateImpl(Attributor &A) override {
+    const Function *F = getAssociatedFunction();
+    if (!F || F->isDeclaration())
+      return indicatePessimisticFixpoint();
+
+    auto NewStates = ArgumentsStates;
+    for (unsigned ArgIdx = 0; ArgIdx < F->arg_size(); ++ArgIdx)
+      if (!isEligibleArgument(*F->getArg(ArgIdx), A, *this))
+        NewStates[ArgIdx] = false;
+
+    bool Changed = NewStates == ArgumentsStates;
+    ArgumentsStates = NewStates;
+    return Changed ? ChangeStatus::CHANGED : ChangeStatus::UNCHANGED;
+  }
+
+  /// See AbstractAttribute::manifest(...).
+  ChangeStatus manifest(Attributor &A) override {
+    Function &F = *getAssociatedFunction();
+    DenseMap<Argument *, Type *> PtrToType;
+    SmallVector<Argument *, 4> CandidateArgs;
+
+    for (unsigned ArgIdx = 0; ArgIdx < F.arg_size(); ++ArgIdx) {
+      Argument *Arg = F.getArg(ArgIdx);
+      if (!isEligibleArgument(*Arg, A, *this))
+        continue;
+
+      CandidateArgs.push_back(Arg);
+      // AAPointerInfo on args
+      for (auto &Use : Arg->uses())
+        if (auto *Store = dyn_cast<StoreInst>(Use.getUser()))
+          PtrToType[Arg] = Store->getValueOperand()->getType();
+    }
+
+    // If there is no valid candidates then return false.
+    if (PtrToType.empty())
+      return indicatePessimisticFixpoint();
+
+    // Create the new struct return type.
+    SmallVector<Type *, 4> OutStructElementsTypes;
+    if (auto *OriginalFuncTy = F.getReturnType(); !OriginalFuncTy->isVoidTy())
+      OutStructElementsTypes.push_back(OriginalFuncTy);
+
+    for (auto *Arg : CandidateArgs)
+      OutStructElementsTypes.push_back(PtrToType[Arg]);
+
+    auto *ReturnStructType = StructType::create(
+        F.getContext(), OutStructElementsTypes, (F.getName() + "_out").str());
+
+    // Get the new Args.
+    SmallVector<Type *, 4> NewParamTypes;
+    for (auto &Arg : F.args())
+      if (!PtrToType.count(&Arg))
+        NewParamTypes.push_back(Arg.getType());
+
+    auto *NewFunctionType =
+        FunctionType::get(ReturnStructType, NewParamTypes, F.isVarArg());
+    auto *NewFunction =
+        Function::Create(NewFunctionType, F.getLinkage(), F.getAddressSpace(),
+                         F.getName() + ".converted");
```

Comment:
Check out 
```
Attributor::isValidFunctionSignatureRewrite
```
I believe it will even check the call sites for you so you don't have to do it yourself (basically ignore what I wrote above).

```
Attributor::registerFunctionSignatureRewrite 
```
And this will do the function cloning and other things for you.

You might need to add a new pair of APIs though:
```
Attributor::isValidFunctionReturnRewrite
```
and
```
Attributor::registerFunctionReturnRewrite
```